In [1]:
import ast
import json
import matplotlib.pyplot as plt
import numpy as np
import os
import pandas as pd
import pickle
import tensorflow as tf
import tensorflow.keras.backend as K
import warnings
warnings.filterwarnings('ignore')

from collections import Counter
from random import randint, random
from scipy.sparse import coo_matrix, hstack
from sklearn.metrics.pairwise import euclidean_distances, cosine_distances, cosine_similarity
from tensorflow import keras
from tqdm import tqdm

## Считаем препроцессные данные из прошлого ноутбучка

In [2]:
interactions_df = pd.read_csv('data_processed/interactions_processed_kion.csv')
users_df = pd.read_csv('data_processed/users_processed_kion.csv')
items_df = pd.read_csv('data_processed/items_processed_kion.csv')

In [3]:
users_df.head()

user_id        age        income sex  kids_flg
0   973171  age_25_34  income_60_90   M      True
1   962099  age_18_24  income_20_40   M     False
2  1047345  age_45_54  income_40_60   F     False
3   721985  age_45_54  income_20_40   F     False
4   704055  age_35_44  income_60_90   F     False

In [4]:
items_df = items_df.rename(columns = {'id' : 'item_id'})

In [5]:
items_df.head()

item_id content_type                 title      title_orig  \
0    10711         film        поговори с ней  Hable con ella   
1     2508         film           голые перцы    Search Party   
2    10716         film      тактическая сила  Tactical Force   
3     7868         film                45 лет        45 Years   
4    16268         film  все решает мгновение            None   

                                             genres       countries  for_kids  \
0           драмы, зарубежные, детективы, мелодрамы         испания     False   
1                  зарубежные, приключения, комедии             сша     False   
2  криминал, зарубежные, триллеры, боевики, комедии          канада     False   
3                      драмы, зарубежные, мелодрамы  великобритания     False   
4                драмы, спорт, советские, мелодрамы            ссср     False   

   age_rating   studios         directors  \
0        16.0   unknown  педро альмодовар   
1        16.0   unknown    скот армстронг   
2        16.0   unknown  адам п. калтраро   
3        16.0   unknown         эндрю хэй   
4        12.0  ленфильм  виктор садовский   

                                              actors  \
0  Адольфо Фернандес, Ана Фернандес, Дарио Гранди...   
1  Адам Палли, Брайан Хаски, Дж.Б. Смув, Джейсон ...   
2  Адриан Холмс, Даррен Шалави, Джерри Вассерман,...   
3  Александра Риддлстон-Барретт, Джеральдин Джейм...   
4  Александр Абдулов, Александр Демьяненко, Алекс...   

                                         description  \
0  Мелодрама легендарного Педро Альмодовара «Пого...   
1  Уморительная современная комедия на популярную...   
2  Профессиональный рестлер Стив Остин («Все или ...   
3  Шарлотта Рэмплинг, Том Кортни, Джеральдин Джей...   
4  Расчетливая чаровница из советского кинохита «...   

                                            keywords release_year_cat  
0  Поговори, ней, 2002, Испания, друзья, любовь, ...        2000-2010  
1  Голые, перцы, 2014, США, друзья, свадьбы, прео...        2010-2020  
2  Тактическая, сила, 2011, Канада, бандиты, ганг...        2010-2020  
3  45, лет, 2015, Великобритания, брак, жизнь, лю...        2010-2020  
4  Все, решает, мгновение, 1978, СССР, сильные, ж...        1970-1980

In [6]:
interactions_df.head()

user_id  item_id last_watch_dt  total_dur  watched_pct
0   176549     9506    2021-05-11       4250           72
1   699317     1659    2021-05-29       8317          100
2   656683     7107    2021-05-09         10            0
3   864613     7638    2021-07-05      14483          100
4   964868     9506    2021-04-30       6725          100

## Готовим фичи пользователей

Посмотрим, какие фичи в датасете фильмов являются категориальными и закодируем их с помощью one-hot encoding.

In [7]:
user_cat_feats = ["age", "income", "sex", "kids_flg"]
# из исходного датафрейма оставим только item_id - этот признак нам понадобится позже
# для того, чтобы маппить айтемы из датафрейма с фильмами с айтемами 
# из датафрейма с взаимодействиями
users_ohe_df = users_df.user_id
for feat in user_cat_feats:
    # получаем датафрейм с one-hot encoding для каждой категориальной фичи
    ohe_feat_df = pd.get_dummies(users_df[feat], prefix=feat)
    # конкатенируем ohe-hot датафрейм с датафреймом, 
    # который мы получили на предыдущем шаге
    users_ohe_df = pd.concat([users_ohe_df, ohe_feat_df], axis=1)

users_ohe_df.head()


user_id  age_age_18_24  age_age_25_34  age_age_35_44  age_age_45_54  \
0   973171              0              1              0              0   
1   962099              1              0              0              0   
2  1047345              0              0              0              1   
3   721985              0              0              0              1   
4   704055              0              0              1              0   

   age_age_55_64  age_age_65_inf  age_age_unknown  income_income_0_20  \
0              0               0                0                   0   
1              0               0                0                   0   
2              0               0                0                   0   
3              0               0                0                   0   
4              0               0                0                   0   

   income_income_150_inf  income_income_20_40  income_income_40_60  \
0                      0                    0                    0   
1                      0                    1                    0   
2                      0                    0                    1   
3                      0                    1                    0   
4                      0                    0                    0   

   income_income_60_90  income_income_90_150  income_income_unknown  sex_F  \
0                    1                     0                      0      0   
1                    0                     0                      0      0   
2                    0                     0                      0      1   
3                    0                     0                      0      1   
4                    1                     0                      0      1   

   sex_M  sex_sex_unknown  kids_flg_False  kids_flg_True  
0      1                0               0              1  
1      1                0               1              0  
2      0                0               1              0  
3      0                0               1              0  
4      0                0               1              0

## Готовим фичи айтемов

Кодируем их точно так же - one-hot'ом.

In [8]:
items_df.head()

item_id content_type                 title      title_orig  \
0    10711         film        поговори с ней  Hable con ella   
1     2508         film           голые перцы    Search Party   
2    10716         film      тактическая сила  Tactical Force   
3     7868         film                45 лет        45 Years   
4    16268         film  все решает мгновение            None   

                                             genres       countries  for_kids  \
0           драмы, зарубежные, детективы, мелодрамы         испания     False   
1                  зарубежные, приключения, комедии             сша     False   
2  криминал, зарубежные, триллеры, боевики, комедии          канада     False   
3                      драмы, зарубежные, мелодрамы  великобритания     False   
4                драмы, спорт, советские, мелодрамы            ссср     False   

   age_rating   studios         directors  \
0        16.0   unknown  педро альмодовар   
1        16.0   unknown    скот армстронг   
2        16.0   unknown  адам п. калтраро   
3        16.0   unknown         эндрю хэй   
4        12.0  ленфильм  виктор садовский   

                                              actors  \
0  Адольфо Фернандес, Ана Фернандес, Дарио Гранди...   
1  Адам Палли, Брайан Хаски, Дж.Б. Смув, Джейсон ...   
2  Адриан Холмс, Даррен Шалави, Джерри Вассерман,...   
3  Александра Риддлстон-Барретт, Джеральдин Джейм...   
4  Александр Абдулов, Александр Демьяненко, Алекс...   

                                         description  \
0  Мелодрама легендарного Педро Альмодовара «Пого...   
1  Уморительная современная комедия на популярную...   
2  Профессиональный рестлер Стив Остин («Все или ...   
3  Шарлотта Рэмплинг, Том Кортни, Джеральдин Джей...   
4  Расчетливая чаровница из советского кинохита «...   

                                            keywords release_year_cat  
0  Поговори, ней, 2002, Испания, друзья, любовь, ...        2000-2010  
1  Голые, перцы, 2014, США, друзья, свадьбы, прео...        2010-2020  
2  Тактическая, сила, 2011, Канада, бандиты, ганг...        2010-2020  
3  45, лет, 2015, Великобритания, брак, жизнь, лю...        2010-2020  
4  Все, решает, мгновение, 1978, СССР, сильные, ж...        1970-1980

In [9]:
item_cat_feats = ['content_type', 'release_year_cat',
                  'for_kids', 'age_rating', 
                  'studios', 'countries', 'directors']

items_ohe_df = items_df.item_id

for feat in item_cat_feats:
    ohe_feat_df = pd.get_dummies(items_df[feat], prefix=feat)
    items_ohe_df = pd.concat([items_ohe_df, ohe_feat_df], axis=1) 

items_ohe_df.head()

item_id  content_type_film  content_type_series  \
0    10711                  1                    0   
1     2508                  1                    0   
2    10716                  1                    0   
3     7868                  1                    0   
4    16268                  1                    0   

   release_year_cat_1920-1930  release_year_cat_1930-1940  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

   release_year_cat_1940-1950  release_year_cat_1950-1960  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           0   

   release_year_cat_1960-1970  release_year_cat_1970-1980  \
0                           0                           0   
1                           0                           0   
2                           0                           0   
3                           0                           0   
4                           0                           1   

   release_year_cat_1980-1990  ...  directors_ярив хоровиц  \
0                           0  ...                       0   
1                           0  ...                       0   
2                           0  ...                       0   
3                           0  ...                       0   
4                           0  ...                       0   

   directors_ярон зильберман  directors_ярополк лапшин  \
0                          0                         0   
1                          0                         0   
2                          0                         0   
3                          0                         0   
4                          0                         0   

   directors_ярослав лупий  directors_ярроу чейни, скотт моужер  \
0                        0                                    0   
1                        0                                    0   
2                        0                                    0   
3                        0                                    0   
4                        0                                    0   

   directors_ясина сезар  directors_ясуоми умэцу  \
0                      0                       0   
1                      0                       0   
2                      0                       0   
3                      0                       0   
4                      0                       0   

   directors_ёдзи фукуяма, ацуко фукусима, николас де креси, синъитиро ватанабэ, сёдзи кавамори  \
0                                                  0                                              
1                                                  0                                              
2                                                  0                                              
3                                                  0                                              
4                                                  0                                              

   directors_ёлкин туйчиев  directors_ён сан-хо  
0                        0                    0  
1                        0                    0  
2                        0                    0  
3                        0                    0  
4                        0                    0  

[5 rows x 8589 columns]

In [10]:
items_df[["description"]].isna().sum()
items_df['description'].fillna("", inplace=True)

## Добавить текстовые признаки в модель. (3 балла)

возьмём берт, через него закодируем description и поместим в датасет как доп фичи

In [11]:
import torch
from tqdm import tqdm
from torch.utils.data import Dataset, DataLoader
from transformers import AutoTokenizer, AutoModel

In [12]:
class TextDataset(Dataset):
    def __init__(self, texts, tokenizer, max_length=2048):
        self.texts = texts
        self.tokenizer = tokenizer
        self.max_length = max_length

    def __len__(self):
        return len(self.texts)

    def __getitem__(self, idx):
        text = self.texts[idx]
        inputs = self.tokenizer(text, max_length=self.max_length, truncation=True, padding='max_length', return_tensors="pt")
        return inputs

In [13]:
device='cuda'

In [14]:
def embed_bert_cls_batch(model, dataloader):
    model.eval()
    all_embeddings = []

    with torch.no_grad():
        for batch in tqdm(dataloader):
            input_ids = batch['input_ids'].squeeze(1).to(device)
            attention_mask = batch['attention_mask'].squeeze(1).to(device)

            outputs = model(input_ids, attention_mask=attention_mask)
            embeddings = outputs.last_hidden_state[:, 0, :]
            embeddings = torch.nn.functional.normalize(embeddings)
            all_embeddings.append(embeddings.cpu().numpy())

    return np.vstack(all_embeddings)

In [15]:
bert_tokenizer = AutoTokenizer.from_pretrained('cointegrated/rubert-tiny2')
bert_model = AutoModel.from_pretrained('cointegrated/rubert-tiny2')

In [16]:
text_dataset = TextDataset(items_df["description"].values, bert_tokenizer)
dataloader = DataLoader(text_dataset, batch_size=16, shuffle=False)

In [17]:
embeddings = embed_bert_cls_batch(bert_model.to(device), dataloader)

100%|████████████████████████████████████████████████████████████████████████████████| 998/998 [06:15<00:00,  2.66it/s]


In [19]:
df_embeddings = pd.DataFrame(embeddings)
df_embeddings.columns = ['description' + str(col) for col in df_embeddings.columns]

In [20]:
items_ohe_df = pd.concat([items_ohe_df, df_embeddings], axis=1)

## Сделаем матрицу взаимодействий

In [21]:
interactions_df.item_id.value_counts()

10440    202457
15297    193123
9728     132865
13865    122119
4151      91167
          ...  
8076          1
8954          1
15664         1
818           1
10542         1
Name: item_id, Length: 15706, dtype: int64

В датасете взаимодействий есть непопулярные фильмы и малоактивные пользователи. Кроме того, в таблице взаимодействий есть события с низким качеством взаимодействия - когда юзер начал смотреть фильм, но вскоре после начала просмотра выключил.

Отфильтруем такие события*, малоактивных юзеров и непопулярные фильмы.

Можете не фильтровать такие события, тогда у вас будет больше негативных примеров.

In [22]:
interactions_df.user_id.value_counts()

416206     1341
1010539     764
555233      685
11526       676
409259      625
           ... 
45493         1
615194        1
96848         1
425823        1
697262        1
Name: user_id, Length: 962179, dtype: int64

In [23]:
print(f"N users before: {interactions_df.user_id.nunique()}")
print(f"N items before: {interactions_df.item_id.nunique()}\n")

# отфильтруем все события взаимодействий, в которых пользователь посмотрел
# фильм менее чем на 10 процентов
interactions_df = interactions_df[interactions_df.watched_pct > 10]

# соберем всех пользователей, которые посмотрели 
# больше 10 фильмов (можете выбрать другой порог)
valid_users = []

c = Counter(interactions_df.user_id)
for user_id, entries in c.most_common():
    if entries > 10:
        valid_users.append(user_id)

# и соберем все фильмы, которые посмотрели больше 10 пользователей
valid_items = []

c = Counter(interactions_df.item_id)
for item_id, entries in c.most_common():
    if entries > 10:
        valid_items.append(item_id)

# отбросим непопулярные фильмы и неактивных юзеров
interactions_df = interactions_df[interactions_df.user_id.isin(valid_users)]
interactions_df = interactions_df[interactions_df.item_id.isin(valid_items)]

print(f"N users after: {interactions_df.user_id.nunique()}")
print(f"N items after: {interactions_df.item_id.nunique()}")

N users before: 962179
N items before: 15706

N users after: 79515
N items after: 6901


После фильтрации может получиться так, что некоторые айтемы/юзеры есть в датасете взаимодействий, но при этом они отсутствуют в датасетах айтемов/юзеров или наоборот. Поэтому найдем id айтемов и id юзеров, которые есть во всех датасетах и оставим только их.

In [24]:
common_users = set(interactions_df.user_id.unique()).intersection(set(users_ohe_df.user_id.unique()))
common_items = set(interactions_df.item_id.unique()).intersection(set(items_ohe_df.item_id.unique()))

print(len(common_users))
print(len(common_items))

interactions_df = interactions_df[interactions_df.item_id.isin(common_items)]
interactions_df = interactions_df[interactions_df.user_id.isin(common_users)]

items_ohe_df = items_ohe_df[items_ohe_df.item_id.isin(common_items)]
users_ohe_df = users_ohe_df[users_ohe_df.user_id.isin(common_users)]

65974
6901


Соберем взаимодействия в матрицу user*item так, чтобы в строках этой матрицы были user_id, в столбцах - item_id, а на пересечениях строк и столбцов - единица, если пользователь взаимодействовал с айтемом и ноль, если нет.

Такую матрицу удобно собирать в numpy array, однако нужно помнить, что numpy array индексируется порядковыми индексами, а нам же удобнее использовать item_id и user_id.

Создадим некие внутренние индексы для user_id и item_id - uid и iid. Для этого просто соберем все user_id и item_id и пронумеруем их по порядку.

In [25]:
interactions_df["uid"] = interactions_df["user_id"].astype("category")
interactions_df["uid"] = interactions_df["uid"].cat.codes

interactions_df["iid"] = interactions_df["item_id"].astype("category")
interactions_df["iid"] = interactions_df["iid"].cat.codes

print(sorted(interactions_df.iid.unique())[:5])
print(sorted(interactions_df.uid.unique())[:5])
interactions_df.head()

[0, 1, 2, 3, 4]
[0, 1, 2, 3, 4]


user_id  item_id last_watch_dt  total_dur  watched_pct    uid   iid
0    176549     9506    2021-05-11       4250           72  10616  3944
1    699317     1659    2021-05-29       8317          100  42131   675
6   1016458      354    2021-08-14       1672           25  61024   139
7    884009      693    2021-08-04        703           14  53150   279
14     5324     8437    2021-04-18       6598           92    310  3485

Отнормируем матрицу взаимодействий

In [26]:
interactions_vec = np.zeros((interactions_df.uid.nunique(), 
                             interactions_df.iid.nunique())) 

for user_id, item_id in zip(interactions_df.uid, interactions_df.iid):
    interactions_vec[user_id, item_id] += 1


res = interactions_vec.sum(axis=1)
for i in range(len(interactions_vec)):
    interactions_vec[i] /= res[i]

In [27]:
print(interactions_df.item_id.nunique())
print(items_ohe_df.item_id.nunique())
print(interactions_df.user_id.nunique())
print(users_ohe_df.user_id.nunique())

print(set(items_ohe_df.item_id.unique()) - set(interactions_df.item_id.unique()))

6897
6901
65974
65974
{11805, 9788, 11501, 1734}


In [28]:
items_ohe_df = items_ohe_df[~items_ohe_df["item_id"].isin([11805, 9788, 11501, 1734])]

Для того, чтобы можно было удобно превратить iid/uid в item_id/user_id и наоборот соберем словари 

{iid: item_id}, {uid: user_id} и {item_id: iid}, {user_id: uid}.

In [29]:
iid_to_item_id = interactions_df[["iid", "item_id"]].drop_duplicates().set_index("iid").to_dict()["item_id"]
item_id_to_iid = interactions_df[["iid", "item_id"]].drop_duplicates().set_index("item_id").to_dict()["iid"]

uid_to_user_id = interactions_df[["uid", "user_id"]].drop_duplicates().set_index("uid").to_dict()["user_id"]
user_id_to_uid = interactions_df[["uid", "user_id"]].drop_duplicates().set_index("user_id").to_dict()["uid"]

И проиндексируем датасеты users_ohe_df и items_ohe_df по внутренним айди:

In [30]:
items_ohe_df["iid"] = items_ohe_df["item_id"].apply(lambda x: item_id_to_iid[x])
items_ohe_df = items_ohe_df.set_index("iid")

users_ohe_df["uid"] = users_ohe_df["user_id"].apply(lambda x: user_id_to_uid[x])
users_ohe_df = users_ohe_df.set_index("uid")

In [31]:
def triplet_loss(y_true, y_pred, n_dims=128, alpha=0.4):
    # будем ожидать, что на вход функции прилетит три сконкатенированных 
    # вектора - вектор юзера и два вектора айтема
    anchor = y_pred[:, 0:n_dims]
    positive = y_pred[:, n_dims:n_dims*2]
    negative = y_pred[:, n_dims*2:n_dims*3]

    # считаем расстояния от вектора юзера до вектора хорошего айтема
    pos_dist = K.sum(K.square(anchor - positive), axis=1)
    # и до плохого
    neg_dist = K.sum(K.square(anchor - negative), axis=1)

    # считаем лосс
    basic_loss = pos_dist - neg_dist + alpha
    loss = K.maximum(basic_loss, 0.0) # возвращаем ноль, если лосс отрицательный
 
    return loss

Попробуйте другие лоссы, например, BPR Triplet loss

## Генератор и семплирование

## Использовать информацию о качестве взаимодействия юзеров с айтемами для более репрезентативного сэмплирования. (3 балла)

позитивный айтем выбирается пропорционально времени просмотра (3 балла), негативный - из непросмотренных

In [33]:
def generator(items, users, interactions, batch_size=1024):
    while True:
        uid_meta = []
        uid_interaction = []
        pos = []
        neg = []
        for _ in range(batch_size):
            # берем рандомный uid
            uid_i = randint(0, interactions.shape[0]-1)
            # id хорошего айтема
            pos_proba = interactions[uid_i]
            pos_proba = pos_proba / pos_proba.sum()  # Нормализуем веса чтобы сумма была равна 1
            pos_i = np.random.choice(range(interactions.shape[1]), p=pos_proba)
            # id плохого айтема
            neg_mask = (interactions[uid_i] == 0)  # Маска для непросмотренных айтемов
            neg_indices = np.arange(interactions.shape[1])[neg_mask]  # Индексы непросмотренных айтемов
            neg_i = np.random.choice(neg_indices)  # Выбор случайного непросмотренного айтема
            # фичи юзера
            uid_meta.append(users.iloc[uid_i])
            # вектор айтемов, с которыми юзер взаимодействовал
            uid_interaction.append(interactions[uid_i])
            # фичи хорошего айтема
            pos.append(items.iloc[pos_i])
            # фичи плохого айтема
            neg.append(items.iloc[neg_i])

        # Преобразуем список в массив numpy перед возвратом
        yield [np.array(uid_meta), np.array(uid_interaction), np.array(pos), np.array(neg)], [np.array(uid_meta), np.array(uid_interaction)]

In [34]:
# инициализируем генератор
gen = generator(items=items_ohe_df.drop(["item_id"], axis=1), 
                users=users_ohe_df.drop(["user_id"], axis=1), 
                interactions=interactions_vec)

ret = next(gen)


print(f"вектор фичей юзера: {ret[0][0].shape}")
print(f"вектор взаимодействий юзера с айтемами: {ret[0][1].shape}")
print(f"вектор 'хорошего' айтема: {ret[0][2].shape}")
print(f"вектор 'плохого' айтема: {ret[0][3].shape}")
print()
print(f"вектор фичей юзера: {ret[1][0].shape}")
print(f"вектор взаимодействий юзера с айтемами: {ret[1][1].shape}")

вектор фичей юзера: (1024, 19)
вектор взаимодействий юзера с айтемами: (1024, 6897)
вектор 'хорошего' айтема: (1024, 8900)
вектор 'плохого' айтема: (1024, 8900)

вектор фичей юзера: (1024, 19)
вектор взаимодействий юзера с айтемами: (1024, 6897)


In [35]:
N_FACTORS = 128

# в датасетах есть столбец user_id/item_id, помним, что он не является фичей для обучения!
ITEM_MODEL_SHAPE = (items_ohe_df.drop(["item_id"], axis=1).shape[1], ) 
USER_META_MODEL_SHAPE = (users_ohe_df.drop(["user_id"], axis=1).shape[1], )

USER_INTERACTION_MODEL_SHAPE = (interactions_vec.shape[1], )

print(f"N_FACTORS: {N_FACTORS}")
print(f"ITEM_MODEL_SHAPE: {ITEM_MODEL_SHAPE}")
print(f"USER_META_MODEL_SHAPE: {USER_META_MODEL_SHAPE}")
print(f"USER_INTERACTION_MODEL_SHAPE: {USER_INTERACTION_MODEL_SHAPE}")

N_FACTORS: 128
ITEM_MODEL_SHAPE: (8900,)
USER_META_MODEL_SHAPE: (19,)
USER_INTERACTION_MODEL_SHAPE: (6897,)


In [36]:
def item_model(n_factors=N_FACTORS):
    # входной слой
    inp = keras.layers.Input(shape=ITEM_MODEL_SHAPE)
    
    # полносвязный слой
    layer_1 = keras.layers.Dense(N_FACTORS, activation='elu', use_bias=False,
                               kernel_regularizer=keras.regularizers.l2(1e-6),
                               activity_regularizer=keras.regularizers.l2(l2=1e-6))(inp)

    # делаем residual connection - складываем два слоя, 
    # чтобы градиенты не затухали во время обучения
    layer_2 = keras.layers.Dense(N_FACTORS, activation='elu', use_bias=False,
                             kernel_regularizer=keras.regularizers.l2(1e-6),
                             activity_regularizer=keras.regularizers.l2(l2=1e-6))(layer_1)
    
    add = keras.layers.Add()([layer_1, layer_2])
    
    # выходной слой
    out = keras.layers.Dense(N_FACTORS, activation='linear', use_bias=False,
                             kernel_regularizer=keras.regularizers.l2(1e-6),
                             activity_regularizer=keras.regularizers.l2(l2=1e-6))(add)
    
    return keras.models.Model(inp, out)


def user_model(n_factors=N_FACTORS):
    # входной слой для вектора фичей юзера (из users_ohe_df)
    inp_meta = keras.layers.Input(shape=USER_META_MODEL_SHAPE)
    # входной слой для вектора просмотров (из iteractions_vec)
    inp_interaction = keras.layers.Input(shape=USER_INTERACTION_MODEL_SHAPE)

    # полносвязный слой
    layer_1_meta = keras.layers.Dense(N_FACTORS, activation='elu', use_bias=False,
                                 kernel_regularizer=keras.regularizers.l2(1e-6),
                                 activity_regularizer=keras.regularizers.l2(l2=1e-6))(inp_meta)

    layer_1_interaction = keras.layers.Dense(N_FACTORS, activation='elu', use_bias=False,
                                 kernel_regularizer=keras.regularizers.l2(1e-6),
                                 activity_regularizer=keras.regularizers.l2(l2=1e-6))(inp_interaction)

    # делаем residual connection - складываем два слоя,
    # чтобы градиенты не затухали во время обучения
    layer_2_meta = keras.layers.Dense(N_FACTORS, activation='elu', use_bias=False,
                                 kernel_regularizer=keras.regularizers.l2(1e-6),
                                 activity_regularizer=keras.regularizers.l2(l2=1e-6))(layer_1_meta)
    

    add = keras.layers.Add()([layer_1_meta, layer_2_meta])
    
    # конкатенируем вектор фичей с вектором просмотров
    concat_meta_interaction = keras.layers.Concatenate()([add, layer_1_interaction])
    
    # выходной слой
    out = keras.layers.Dense(N_FACTORS, activation='linear', use_bias=False,
                             kernel_regularizer=keras.regularizers.l2(1e-6),
                             activity_regularizer=keras.regularizers.l2(l2=1e-6))(concat_meta_interaction)
    
    return keras.models.Model([inp_meta, inp_interaction], out)

# инициализируем модели юзера и айтема
i2v = item_model()
u2v = user_model()

# вход для вектора фичей юзера (из users_ohe_df)
ancor_meta_in = keras.layers.Input(shape=USER_META_MODEL_SHAPE)
# вход для вектора просмотра юзера (из interactions_vec)
ancor_interaction_in = keras.layers.Input(shape=USER_INTERACTION_MODEL_SHAPE)

# вход для вектора "хорошего" айтема
pos_in = keras.layers.Input(shape=ITEM_MODEL_SHAPE)
# вход для вектора "плохого" айтема
neg_in = keras.layers.Input(shape=ITEM_MODEL_SHAPE)

# получаем вектор юзера
ancor = u2v([ancor_meta_in, ancor_interaction_in])
# получаем вектор "хорошего" айтема
pos = i2v(pos_in)
# получаем вектор "плохого" айтема
neg = i2v(neg_in)

# конкатенируем полученные векторы
res = keras.layers.Concatenate(name="concat_ancor_pos_neg")([ancor, pos, neg])

# собираем модель
model = keras.models.Model([ancor_meta_in, ancor_interaction_in, pos_in, neg_in], res)

### Попробуйте передать текстовые признаки в модель

In [37]:
model_name = 'recsys_resnet_linear'

# логируем процесс обучения в тензорборд
t_board = keras.callbacks.TensorBoard(log_dir=f'runs/{model_name}')

# уменьшаем learning_rate, если лосс долго не уменьшается (в течение двух эпох)
decay = keras.callbacks.ReduceLROnPlateau(monitor='loss', patience=2, factor=0.8, verbose=1)

# сохраняем модель после каждой эпохи, если лосс уменьшился
check = keras.callbacks.ModelCheckpoint(filepath=model_name + '/epoch{epoch}-{loss:.2f}.h5', monitor="loss")


In [38]:
# компилируем модель, используем оптимайзер Adam и triplet loss
opt = keras.optimizers.Adam(lr=0.001)
model.compile(loss=triplet_loss, optimizer=opt)

In [39]:
# модель айтема
item_model().summary()

Model: "model_3"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_8 (InputLayer)           [(None, 8900)]       0           []                               
                                                                                                  
 dense_7 (Dense)                (None, 128)          1139200     ['input_8[0][0]']                
                                                                                                  
 dense_8 (Dense)                (None, 128)          16384       ['dense_7[0][0]']                
                                                                                                  
 add_2 (Add)                    (None, 128)          0           ['dense_7[0][0]',                
                                                                  'dense_8[0][0]']          

In [40]:
# модель юзера
user_model().summary()

Model: "model_4"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_9 (InputLayer)           [(None, 19)]         0           []                               
                                                                                                  
 dense_10 (Dense)               (None, 128)          2432        ['input_9[0][0]']                
                                                                                                  
 dense_12 (Dense)               (None, 128)          16384       ['dense_10[0][0]']               
                                                                                                  
 input_10 (InputLayer)          [(None, 6897)]       0           []                               
                                                                                            

In [41]:
# общая модель
model.summary()

Model: "model_2"
__________________________________________________________________________________________________
 Layer (type)                   Output Shape         Param #     Connected to                     
 input_4 (InputLayer)           [(None, 19)]         0           []                               
                                                                                                  
 input_5 (InputLayer)           [(None, 6897)]       0           []                               
                                                                                                  
 input_6 (InputLayer)           [(None, 8900)]       0           []                               
                                                                                                  
 input_7 (InputLayer)           [(None, 8900)]       0           []                               
                                                                                            

In [42]:
# начинаем обучение, не забывая дропнуть столбцы item_id и user_id 
# из датафреймов при инициализации генератора.

# batch_size можно (и лучше) поставить побольше, если вы не органичены в ресурсах

model.fit(generator(items=items_ohe_df.drop(["item_id"], axis=1), 
                    users=users_ohe_df.drop(["user_id"], axis=1), 
                    interactions=interactions_vec,
                    batch_size=64), 
          steps_per_epoch=100, 
          epochs=2, 
          initial_epoch=0,
          callbacks=[decay, t_board, check]
)

Epoch 1/2
100/100 [==============================] - 108s 1s/step - loss: 0.3145 - lr: 0.0010
Epoch 2/2
100/100 [==============================] - 105s 1s/step - loss: 0.2172 - lr: 0.0010


In [43]:
# берем рандомного юзера
rand_uid = np.random.choice(list(users_ohe_df.index))

# получаем фичи юзера и вектор его просмотров айтемов
user_meta_feats = users_ohe_df.drop(["user_id"], axis=1).iloc[rand_uid]
user_interaction_vec = interactions_vec[rand_uid]

# берем рандомный айтем
rand_iid = np.random.choice(list(items_ohe_df.index))
# получаем фичи айтема
item_feats = items_ohe_df.drop(["item_id"], axis=1).iloc[rand_iid]

# получаем вектор юзера
user_vec = u2v.predict([np.array(user_meta_feats).reshape(1, -1), 
                        np.array(user_interaction_vec).reshape(1, -1)])

# и вектор айтема
item_vec = i2v.predict(np.array(item_feats).reshape(1, -1))

# считаем расстояние между вектором юзера и вектором айтема
from sklearn.metrics.pairwise import euclidean_distances as ED

ED(user_vec, item_vec)

array([[1.7638429]], dtype=float32)

In [44]:
# получаем фичи всех айтемов
items_feats = items_ohe_df.drop(["item_id"], axis=1).to_numpy()
# получаем векторы всех айтемов
items_vecs = i2v.predict(items_feats)

# считаем расстояния
dists = ED(user_vec, items_vecs)

In [45]:
items_vecs.shape

(6897, 128)

In [46]:
users_meta_feats = users_ohe_df.drop(["user_id"], axis=1)
users_interaction_vec = interactions_vec

In [47]:
users_meta_feats.shape

(65974, 19)

In [48]:
users_interaction_vec.shape

(65974, 6897)

In [49]:
np.array(users_meta_feats).shape

(65974, 19)

In [50]:
users_vec = u2v.predict([np.array(users_meta_feats), 
                        np.array(users_interaction_vec)])

In [51]:
users_vec.shape

(65974, 128)

In [52]:
items_vecs.shape

(6897, 128)

In [53]:
dists = ED(users_vec, items_vecs)

In [54]:
dists.shape

(65974, 6897)

In [55]:
top10_iids = np.argsort(dists, axis=1)[:,:10]

In [56]:
top10_iids.reshape(dists.shape[0], 10).shape

(65974, 10)

In [57]:
top10_iids

array([[ 649, 1967, 3574, ...,  155, 5008, 2795],
       [6403, 1533, 5410, ..., 6482, 3574, 1067],
       [3574,  693, 5858, ..., 5802, 2236, 1580],
       ...,
       [ 649, 3574,  572, ..., 4061, 5858, 5342],
       [3574,  572,   53, ..., 4664, 5017,  693],
       [3868, 2629,  693, ..., 2924, 5136, 1533]], dtype=int64)

In [58]:
top10_iids.shape

(65974, 10)

In [59]:
top10_iids_item = [iid_to_item_id[iid] for iid in top10_iids.reshape(-1)]

In [60]:
top10_iids_item = np.array(top10_iids_item).reshape(top10_iids.shape)

In [61]:
top10_iids_item.shape

(65974, 10)

In [62]:
df_dssm = pd.DataFrame(columns = ['user_id', 'item_id'])

In [64]:
df_dssm = pd.DataFrame({'user_id': [uid_to_user_id[uid] for uid in np.arange(top10_iids_item.shape[0])]})

In [65]:
df_dssm['item_id'] = list(top10_iids_item)

In [66]:
df_dssm

user_id                                            item_id
0            2  [1590, 4731, 8636, 3935, 7582, 12173, 11131, 3...
1           21  [15297, 3734, 12995, 12173, 142, 12080, 1699, ...
2           53  [8636, 1699, 14053, 7425, 6210, 16270, 2301, 1...
3           60  [14053, 16270, 1590, 1699, 8636, 11109, 6404, ...
4           81  [12080, 8636, 12841, 7582, 1590, 391, 14053, 1...
...        ...                                                ...
65969  1097486  [12995, 3734, 2647, 3669, 12080, 15297, 7102, ...
65970  1097489  [14053, 1699, 10605, 8636, 1590, 2237, 7417, 7...
65971  1097508  [1590, 8636, 1418, 11109, 7582, 3935, 391, 981...
65972  1097513  [8636, 1418, 142, 1650, 3935, 15297, 15464, 11...
65973  1097516  [9342, 6404, 1699, 12995, 10605, 1692, 11310, ...

[65974 rows x 2 columns]

In [67]:
df_dssm = df_dssm.explode('item_id')
df_dssm['rank'] = df_dssm.groupby('user_id').cumcount() + 1
df_dssm = df_dssm.groupby('user_id').agg({'item_id': list}).reset_index()

In [68]:
df_dssm.head()

user_id                                            item_id
0        2  [1590, 4731, 8636, 3935, 7582, 12173, 11131, 3...
1       21  [15297, 3734, 12995, 12173, 142, 12080, 1699, ...
2       53  [8636, 1699, 14053, 7425, 6210, 16270, 2301, 1...
3       60  [14053, 16270, 1590, 1699, 8636, 11109, 6404, ...
4       81  [12080, 8636, 12841, 7582, 1590, 391, 14053, 1...

In [69]:
df_dssm

user_id                                            item_id
0            2  [1590, 4731, 8636, 3935, 7582, 12173, 11131, 3...
1           21  [15297, 3734, 12995, 12173, 142, 12080, 1699, ...
2           53  [8636, 1699, 14053, 7425, 6210, 16270, 2301, 1...
3           60  [14053, 16270, 1590, 1699, 8636, 11109, 6404, ...
4           81  [12080, 8636, 12841, 7582, 1590, 391, 14053, 1...
...        ...                                                ...
65969  1097486  [12995, 3734, 2647, 3669, 12080, 15297, 7102, ...
65970  1097489  [14053, 1699, 10605, 8636, 1590, 2237, 7417, 7...
65971  1097508  [1590, 8636, 1418, 11109, 7582, 3935, 391, 981...
65972  1097513  [8636, 1418, 142, 1650, 3935, 15297, 15464, 11...
65973  1097516  [9342, 6404, 1699, 12995, 10605, 1692, 11310, ...

[65974 rows x 2 columns]

In [71]:
df_dssm.to_csv('dssm_offline.csv', index=False)

## Пробить скор бейзлайна DSSM модели 0.075. (2 балла)

На лб скор очень близкий к популярному, т.к. предсказания только для 65974 юзеров, остальное - популярные. Метрику в 0.075 пробивает с запасом

## Обернуть в сервис, который должен  критериям читаемости и воспоизводимости. (5 балла) 

Сделал оффлайн